### DISCUSSION TITLE
Persist structured, continous data for Tauri Android application built on wasm

### DISCUSSION THREAD

**Discussion by opeolluwa** [tauri-apps/tauri#12051] [Q&A]
Hello here, 

What is the best way to persist structured data on Mobile, 
I'm using Leptos and Tauri V2.

I tried bundling [rusqlite](https://crates.io/crates/rusqlite), [sqlite-web](https://github.com/xmtp/sqlite-web-rs) and [gluesql](https://gluesql.org/docs/0.16.0/) crates.

So far, only gluesql successfully integrates without a build error, however it's not so satisfactory. I'd love to use sqlite or of sort

**Comment by FabianLars** [comment]
rusqlite should work as well. We use `sqlx` in our sql plugin and that works fine on mobile (sqlx uses rusqlite's libsqlite3-sys crate).

**Reply by opeolluwa** [reply]
Thank you so much @FabianLars I had previously tried it to no avail. 

  
```toml
[dependencies]
...
rusqlite = { version = "0.32.0", features = ["bundled"] }
```

it gives this error `cargo:warning=sqlite3/sqlite3.c:14678:10: fatal error: 'stdio.h' file not found`  when I run `cargo build --lib --target "wasm32-unknown-unknown"` 

I appreciate your help, and happy holidays!

**Reply by FabianLars** [reply]
Ohhh, now i understand the title. I don't think any sqlite library can be compiled to wasm. But even _if_ one can, it would only support in-memory databases because your frontend doesn't have direct access to the file system.
You'll have to do the database stuff on the "backend"/ the tauri side.

**Reply by opeolluwa** [reply]
Yes! 

I figured that out too already, and I was able to get it to work for desktop target (macOS) but not for the Android platform 

But I was getting `test.db` could not be found 

Where test.db is my sqlite database file, 

I tried bundling it as Tauri resource  https://v2.tauri.app/develop/resources/

I also tried std::fs and the include_dir crate, it all resulted to the same error

**Reply by FabianLars** [reply]
paths relative to the cwd probably just can't work. in the sql plugin we make it relative to the app_data_dir or app_config_dir (forgot which one) and that works.

**Reply by opeolluwa** [reply]
Thank you so much @FabianLars I'm going to check this out

**Comment by opeolluwa** [comment]
## Folder structure 
<img width="187" alt="Screenshot 2024-12-29 at 18 04 51" src="https://github.com/user-attachments/assets/37a18622-7280-4723-9c75-af31e3c0590e" />

## Code 

### The database crate
```rust
// sqlite_wasm_bindgen/lib.rs

pub struct SqliteWasm {}

#[allow(dead_code)]
impl SqliteWasm {
    pub fn init(database_path: &str) -> rusqlite::Result<Connection> {
        let database_connection = Connection::open(database_path)?;
        // let database_connection = Connection::open_in_memory()?;

        let create_tables_result =
            database_connection.execute_batch(&Self::create_table_statements());

       ...
        }
    }
```

### Tauri/backend crate
```rust
// tauri/lib.rs
#[cfg_attr(mobile, tauri::mobile_entry_point)]
pub fn run() {
    tauri::Builder::default()
        .setup(|app| {
            let database_path = "test.db";
            let connection = sqlite_wasm_bindgen_database::SqliteWasm::init(&database_path)
                .expect("error initializing database");

            app.manage(state::AppState {
                conn: Mutex::new(connection),
            });
            Ok(())
        })
       ...
        .expect("error while running tauri application");
}
```

### Error
```sh
12-29 18:28:37.344  5451  5519 I RustStdoutStderr: thread '<unnamed>' panicked at tauri/src/lib.rs:32:18:
12-29 18:28:37.345  5451  5519 I RustStdoutStderr: error initializing database: SqliteFailure(Error { code: CannotOpen, extended_code: 14 }, Some("unable to open database file: test.db"))
```
<img width="1162" alt="Screenshot 2024-12-29 at 18 30 01" src="https://github.com/user-attachments/assets/7c8b82db-2617-4c0c-b992-f4a4003d4536" />
